lib/AST/SearchPathOptions.cpp

```diff
@@ -0,0 +1,96 @@
+//===----------------------- SearchPathOptions.cpp ------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/AST/SearchPathOptions.h"
+#include "llvm/ADT/SmallSet.h"
+
+using namespace swift;
+
+void ModuleSearchPathLookup::addFilesInPathToLookupTable(
+    llvm::vfs::FileSystem *FS, StringRef SearchPath, ModuleSearchPathKind Kind,
+    bool IsSystem, unsigned SearchPathIndex) {
+  std::error_code Error;
+  for (auto Dir = FS->dir_begin(SearchPath, Error);
+       !Error && Dir != llvm::vfs::directory_iterator(); Dir.increment(Error)) {
+    StringRef Filename = llvm::sys::path::filename(Dir->path());
+    LookupTable[Filename].push_back(
+        {SearchPath.str(), Kind, IsSystem, SearchPathIndex});
```

Comment:
If we store the results of `getDarwinImplicitFrameworkSearchPaths ` in `SearchPathOptions` this could then be a `StringRef` (I think). Possibly worth it?

Reply:
With the current design, we only decide whether we need these paths (i.e. whether we are on a Darwin platform) when enumerating the search paths. So if we want to store the Darwin implicit framework search paths in `SearchPathOptions`, we would need to always compute them when setting SDK path, regardless of the platform (or have a slightly more complicated caching method). I’m not sure if it’s worth it. My current gut feeling is that this function will be called <10 times per compilation/SourceKit operation.

The clean solution would be that `SearchPathOptions` knows whether it’s on a Darwin platform, but I think that’s out of scope for this PR.

Reply:
Just seemed like a relatively large map to me, but I guess there's not *that* many file -> search paths. I was mostly hoping we could move the platform check into `ParseSearchPathArgs` and just add the extra paths directly to the framework search paths rather than have separate field, but I haven't looked into that at all. If that's not possible then I don't really have a preference either way 🤷

Reply:
Ah, sorry. I misunderstood you. I though you wanted to avoid the computation of `getDarwinImplicitFrameworkSearchPaths`. Making `ModuleSearchPath` only contain a `StringRef` does seem like a worthwhile improvement 👍 

> I was mostly hoping we could move the platform check into ParseSearchPathArgs and just add the extra paths directly to the framework search paths rather than have separate field, but I haven't looked into that at all. If that's not possible then I don't really have a preference either way 🤷

We can’t do that because framework search paths may be added after the compiler invocation is created and those should come before the Darwin implicit search paths.

Reply:
Ah I see. Yeah the computation didn't really bother me at all - from a quick look this just seemed to be the only part that was preventing using StringRefs.